network/crl_test.go

```diff
@@ -0,0 +1,392 @@
+	}
+	data, err = cache.Get("test1")
+	if data != nil {
+		t.Fatalf("Unexpected data while reading removed cache item")
+	}
+	if err == nil {
+		t.Fatalf("Unexpected error while reading removed cache item")
+	}
+}
+
+func TestLRUParsedCRLCache(t *testing.T) {
+	// Same as TestDefaultCRLCache, but with *pkix.CertificateList instead of []byte as value
+	cache := NewLRUParsedCRLCache(4)
+
+	crl, err := x509.ParseCRL(getTestCRL())
+	if err != nil {
+		t.Fatal("Failed to parse test CRL")
+	}
+
+	// we don't expect to see any items in cache when it's created
+	cachedCRL, err := cache.Get("test1")
+	if cachedCRL != nil {
+		t.Fatalf("Unexpected data while reading empty cache")
+	}
+	if err == nil {
+		t.Fatalf("No expected error while reading empty cache")
+	}
+
+	// let's insert something
+	cache.Put("test1", crl)
+	cachedCRL, err = cache.Get("test1")
+	if cachedCRL == nil {
+		t.Fatalf("Unexpected fail while reading recently inserted cache item")
+	}
+	if err != nil {
+		t.Fatalf("Unexpected error while reading recently inserted cache item")
+	}
+
+	// and test removal
+	err = cache.Remove("test1")
+	if err != nil {
+		t.Fatalf("Unexpected error while removing recently inserted cache item")
+	}
+	cachedCRL, err = cache.Get("test1")
+	if cachedCRL != nil {
+		t.Fatalf("Unexpected data while reading removed cache item")
+	}
+	if err == nil {
+		t.Fatalf("Unexpected error while reading removed cache item")
+	}
+}
+
+func TestDefaultCRLVerifier(t *testing.T) {
+	crlConfig := CRLConfig{uri: "http://127.0.0.1:8889/crl.pem", fromCert: crlFromCertIgnore}
+	crlVerifier := DefaultCRLVerifier{Config: crlConfig, Client: DefaultCRLClient{}, Cache: &DefaultCRLCache{}, ParsedCache: NewLRUParsedCRLCache(16)}
+
+	// Fool crlVerifier into thinking the CRL is already in cache to avoid performing requests.
+	// CRLCache and CRLClient are tested separately anyway.
+	crlVerifier.Cache.Put("http://127.0.0.1:8889/crl.pem", getTestCRL())
+
+	//
+	// Test valid certificate chain
+	//
+	validRawCerts := [][]byte{}
+	validRawCerts = append(validRawCerts, getTestValidCert())
+	validRawCerts = append(validRawCerts, getTestCACert())
+
+	validVerifiedChains := [][]*x509.Certificate{}
+	validVerifiedChains = append(validVerifiedChains, []*x509.Certificate{})
+	cert, err := pemToX509Cert(validRawCerts[0])
+	if err != nil {
+		t.Fatalf("Failed to parse test PEM certificate: %v", err)
+	}
+	validVerifiedChains[0] = append(validVerifiedChains[0], cert)
+	cert, err = pemToX509Cert(validRawCerts[1])
+	if err != nil {
+		t.Fatalf("Failed to parse test PEM certificate: %v", err)
+	}
+	validVerifiedChains[0] = append(validVerifiedChains[0], cert)
+
+	t.Logf("validVerifiedChains[0][0].Subject = %s", validVerifiedChains[0][0].Subject.String())
+	t.Logf("validVerifiedChains[0][1].Subject = %s", validVerifiedChains[0][1].Subject.String())
+	if err = validVerifiedChains[0][0].CheckSignatureFrom(validVerifiedChains[0][1]); err != nil {
```

Comment:
what we test here?) in this testcase we should test our DefaultCRLVerifier and be sure that it use valid test mixtures. If we need check test data, better to do it in separate function (or better check it once manually to be sure that it is valid and it's enough for that. it can't change their state after that from valid to invalid ^ )

Reply:
It's more like a small sanity check to make sure the first one is issued by the second one